sklearn/neighbors/tests/test_graph.py

```diff
@@ -0,0 +1,75 @@
+import numpy as np
+
+from sklearn.metrics import euclidean_distances
+from sklearn.neighbors import KNeighborsTransformer, RadiusNeighborsTransformer
+
+
+def test_transformer_nonzero():
+    # Test the number of neighbors returned
+    n_neighbors = 5
+    n_samples_fit = 20
+    n_samples_transform = 18
+    n_features = 10
+
+    rng = np.random.RandomState(42)
+    X = rng.randn(n_samples_fit, n_features)
+    X2 = rng.randn(n_samples_transform, n_features)
+    radius = np.percentile(euclidean_distances(X), 10)
+
+    # with n_neighbors
+    for mode in ('distance', 'connectivity'):
+        nnt = KNeighborsTransformer(n_neighbors=n_neighbors, mode=mode)
+        Xt = nnt.fit_transform(X)
+        assert Xt.shape == (n_samples_fit, n_samples_fit)
+        assert Xt.data.shape == (n_samples_fit * n_neighbors, )
+
+        X2t = nnt.transform(X2)
+        assert X2t.shape == (n_samples_transform, n_samples_fit)
+        assert X2t.data.shape == (n_samples_transform * n_neighbors, )
+
+    # with radius
+    for mode in ('distance', 'connectivity'):
+        nnt = RadiusNeighborsTransformer(radius=radius, mode=mode)
+        Xt = nnt.fit_transform(X)
+        assert Xt.shape == (n_samples_fit, n_samples_fit)
+        assert not Xt.data.shape == (n_samples_fit * n_neighbors, )
+
+        X2t = nnt.transform(X2)
+        assert X2t.shape == (n_samples_transform, n_samples_fit)
+        assert not X2t.data.shape == (n_samples_transform * n_neighbors, )
+
+
+def test_include_self_logic():
+    # Test the effect of parameter include_self
+    n_neighbors = 5
+    n_samples_fit, n_samples_transform, n_features = 20, 18, 10
+    rng = np.random.RandomState(42)
+    X = rng.randn(n_samples_fit, n_features)
+    X2 = rng.randn(n_samples_transform, n_features)
+
+    # Same behavior for both cases
+    for include_self in [True, False]:
+        nnt = KNeighborsTransformer(n_neighbors=n_neighbors,
+                                    include_self=include_self)
+        Xt = nnt.fit_transform(X)
+        # Each sample is it's own neighbor
+        assert np.all(Xt.diagonal() == 0)
```

Comment:
This is going to be true whether or not the diagonal is explicit. To check for explicit vs implicit zeros on diagonal, easiest way might be:
```py
def _get_explicit_diagonals(X):
    """Returns indices of rows where the diagonal is explicitly stored"""
    X = X.tocoo()
    return X.row[X.row == X.col]
```